README.md
# g-mapify [![NPM version](https://img.shields.io/npm/v/g-mapify.svg)](https://www.npmjs.com/package/g-mapify) [![Downloads](http://img.shields.io/npm/dm/g-mapify.svg)](https://npmjs.org/package/g-mapify) [![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg?style=flat-square)](#contributors-)
Flexible react google map with more options of search, pick & select.

![g-mapify](https://raw.githubusercontent.com/bharatpe/g-mapify/master/g-mapify.png)


![g-mapify](https://raw.githubusercontent.com/bharatpe/g-mapify/master/example/images/g-mapify.png)


# Getting Started

```javascript
yarn add g-mapify
```

or

```javascript
npm i g-mapify
```


## Usage
Basic use:
```javascript
    import  React from  'react';
    import { GMapify } from  'g-mapify';
    import  'g-mapify/dist/index.css';
    
    const  App = () => {
	    return (
		    <GMapify appKey="[google-map-key]" />
	    )
    }
```

### Example: Basic map
![basic map image](https://raw.githubusercontent.com/bharatpe/g-mapify/master/example/images/basicmap.png?raw=true)


## Example with options (Search Map)
```javascript
import  React from  'react';
import { GMapify } from  'g-mapify';
import  'g-mapify/dist/index.css';

const  App = () => {
    const  onMapSelect = (status, data) => console.warn(status, data);
    
    return (
        <GMapify appKey="[google-map-key]" hasSearch onSelect={onMapSelect}/>
    )
}
```
 #### Note
 If *status* is getting *false* that means something happened bad on location search.

![Map Search image](https://raw.githubusercontent.com/bharatpe/g-mapify/master/example/images/search-map-desktop.png?raw=true)

## Example with multiple markers
```javascript
import  React from  'react';
import { GMapify } from  'g-mapify';
import  'g-mapify/dist/index.css';

const  App = () => {

    const markers = [
        [28.4165425, 77.0437857, "Hello 1"],
        [28.415671, 77.0520993, `<div><h3>Southcity 2</h3> <img src="https://www.rentomojo.com/blog/wp-content/uploads/2019/07/shutterstock_1298400742.jpg" height="140"/> </div>`],
        [28.4175717, 77.05284840000002]
    ]

    const  onMapSelect = (status, data) => console.warn(status, data);
    
    return (
        <GMapify 
        appKey="[google-map-key]" 
        customMarkers={markers}
        autoCenter={false}
        onSelect={onMapSelect}/>
    )
}
```

![Map Mutiple markers image](https://raw.githubusercontent.com/bharatpe/g-mapify/master/example/images/markers-map.png?raw=true)


## Options
|  Props                        | Description                                           | Default value
|-------------------------------|-------------------------------------------------------|-------------------------------|
|appKey (Mandatory)             | google map app key                                    | It's a mandatory value
|lat                            | default latitude of map                               | 28.7041
|lng                            | default logitude of map                               | 77.1025
|mapOptions                     | google map options that you can change or add more    | zoom:  15, zoomControl:  false, mapTypeControl:  false, fullscreenControl:  false, streetViewControl:  false, clickableIcons:  false, mapTypeId:  "roadmap"
|libraries                      | add more funcationality on google map                 | places
|mapClassName                   | custom class name for map element                     | None
|hasMarker                      | show marker in map                                    | true
|markerIcon                     | map marker icon                                       | Blue marker icon
|autoCenter                     | Auto center map on move or zoom                       | true
|customMarkers                  | Multiple custom markers to mark on map                | [] (Example :  [[lat, lng, content], [lat, lng, content], ...] )
|hasSearch                      | apply search in map                                   | false
|mapSearchPlace                 | map search input box place (required a valid selector name like (#id, .class)) | bottom of the map
|debounceTime                   | search debounce time (unit *ms*)                      | 2000
|inputClassName                 | custom class name for Search input                    | None
|searchPlaceHolder              | placeholder of search box                             | Search here
|searchClassName                | custom class name of search container                 | None
|onSelect                       | trigger whenever map position changed (return *status: [true/false], data: [map data object], mapStatus: [map native status]*)| 

#### Note 
When use customMarkers (More than 1) - Make sure autoCenter, hasSearch are set to false. Which dosen't make sense any way with multiple markers.


## Address Formatter
You can get City, Pincode, State etc. from *address_components* using **addressFormatter** method. This method transfrom *address_components* array data into *object*.
```javascript
import { GMapify, AddressFormatter } from  'g-mapify'

const  onMapSelect = (status, data) => {
    if (status) {
        console.warn('Map Data', data);
        
        // get formatted address from google map address_components
        const  formattedAddress = AddressFormatter(data.address_components);
        console.warn('formated address', formattedAddress);
    }
};
```

## Refer example for all 3 map variations

![Map variations image](https://raw.githubusercontent.com/bharatpe/g-mapify/master/example/images/example.png?raw=true)



# Releases
Here [Releases](https://github.com/bharatpe/g-mapify/blob/master/releases.md)

    
## Note: For Contributors
This repo build with [create-react-library](https://github.com/transitive-bullshit/create-react-library)


## Contributors
Here [Contributors](https://github.com/bharatpe/g-mapify/graphs/contributors)


## License
ISC [BharatPe TM](https://github.com/BharatPe)


releases.md
# Releases

[Releases](https://github.com/bharatpe/g-mapify/releases)

example/src/App.js
import React, { useCallback, useState, useRef } from 'react'
import { GMapify, AddressFormatter } from 'g-mapify'
import 'g-mapify/dist/index.css'
import './index.css'
// import InputDefault from './components/Input';

const BASIC = 'basic_map';
const SEARCH_MAP = 'search_map';
const MULTIPLE_MARKER = 'marker_map';

const App = () => {
  const mapOptions = {
    zoom: 15
  };

  const markers = [
    [28.4165425, 77.0437857, "Hello 1"],
    [28.415671, 77.0520993, `<div><h3>Southcity 2</h3> <img src="https://www.rentomojo.com/blog/wp-content/uploads/2019/07/shutterstock_1298400742.jpg" height="140"/> </div>`],
    [28.4175717, 77.05284840000002, "<h1> Location </h1>"]
  ]

  const [mapType, setmapType] = useState(BASIC);

  const onMapSelect = (status, data) => {
    console.warn('Map Data', data);

    // get formatted address from google map address_components
    const formattedAddress = AddressFormatter(data.address_components);
    console.warn('formated address', formattedAddress);
  };

  return (
  <>
    <button onClick={() => setmapType(BASIC)} disabled={mapType === BASIC}>Basic Map</button>
    <button onClick={() => setmapType(SEARCH_MAP)} disabled={mapType === SEARCH_MAP}>Search Map</button>
    <button onClick={() => setmapType(MULTIPLE_MARKER)} disabled={mapType === MULTIPLE_MARKER}>Custom Map</button>
    
    { 
      mapType === BASIC && <GMapify key={1} mapOptions={mapOptions} appKey="AIzaSyBeZ5-CnQtbhPaEr2u162G-SoCF44lQRAg" mapClassName="h-100" />
    }
    { 
      mapType === SEARCH_MAP && <GMapify key={2} mapOptions={mapOptions} appKey="AIzaSyBeZ5-CnQtbhPaEr2u162G-SoCF44lQRAg" mapClassName="h-100" hasSearch onSelect={onMapSelect} />
    }
    { 
      mapType === MULTIPLE_MARKER && <GMapify key={3} mapOptions={mapOptions} appKey="AIzaSyBeZ5-CnQtbhPaEr2u162G-SoCF44lQRAg" mapClassName="h-100" autoCenter={false} customMarkers={markers} onSelect={onMapSelect} />
    }
  </>
  )
}

export default App
